db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/FromBaseTable.java

```diff
@@ -1952,43 +1980,61 @@ private ResultColumnList newResultColumns(
             boolean cloneRCs)
             throws StandardException{
         IndexRowGenerator irg=idxCD.getIndexDescriptor();
-        int[] baseCols=irg.baseColumnPositions();
-        ResultColumnList newCols=
-                (ResultColumnList)getNodeFactory().getNode(
+        ResultColumnList newCols =
+                (ResultColumnList) getNodeFactory().getNode(
                         C_NodeTypes.RESULT_COLUMN_LIST,
                         getContextManager());
 
-        for(int basePosition : baseCols){
-            ResultColumn oldCol=oldColumns.getResultColumn(basePosition);
-            ResultColumn newCol;
+        if (irg.isOnExpression()) {
+            // When building new ResultColumn instances, we don't need to set expression
+            // or virtual column number as they are not needed. In case of a scan on an
+            // expression-based index, these are just placeholders. All we care about is
+            // that they should all be referenced for now so that we can build the
+            // template row, then we will clear reference status and set them properly.
```

Comment:
This comment is true only for this change because all expression-based indexes are non-covering indexes. In next step, to rewrite an index expression to a column reference, we do the following:
```
rc.setIndexExpression(exprAst);
rc.setReferenced();
rc.setVirtualColumnId(i + 1);  // virtual column IDs are 1-based
rc.setName(idxCD.getConglomerateName() + "_col" + rc.getColumnPosition());
```